transports/onion/src/lib.rs

```diff
@@ -0,0 +1,123 @@
+#![doc(html_logo_url = "https://libp2p.io/img/logo_small.png")]
+#![doc(html_favicon_url = "https://libp2p.io/img/favicon.png")]
+
+use std::pin::Pin;
+use std::sync::Arc;
+
+use address::dangerous_extract_tor_address;
+use arti_client::{TorAddrError, TorClient, TorClientBuilder};
+use futures::{future::BoxFuture, FutureExt};
+use libp2p_core::{transport::TransportError, Multiaddr, Transport};
+use tor_rtcompat::PreferredRuntime;
+
+mod address;
+mod provider;
+
+#[doc(inline)]
+pub use provider::OnionStream;
+
+#[derive(Debug, thiserror::Error)]
+pub enum OnionError {
+    #[error("error during address translation")]
+    AddrErr(#[from] TorAddrError),
+    #[error("error in arti")]
+    ArtiErr(#[from] arti_client::Error),
+    #[error("onion services are not implented yet, since arti doesn't support it. (awaiting Arti 1.2.0)")]
+    OnionServiceUnimplemented,
+}
+
+#[derive(Clone)]
+pub struct OnionClient {
+    // client is in an Arc, because wihtout it the Transport::Dial method can't be implemented,
+    // due to lifetime issues. With the, eventual, stabilization of static async traits this issue
+    // will be resolved.
+    client: Arc<TorClient<PreferredRuntime>>,
+}
+
+pub type OnionBuilder = TorClientBuilder<PreferredRuntime>;
```

Comment:
This is leaking the `arti_client` dependency to our consumers, meaning it affects semver when we upgrade the `arti_client` dependency.

I think that is okay for now.